doc/design/service-discovery.md

```diff
@@ -0,0 +1,228 @@
+# Service Discovery
+
+## Summary and problem description
+
+Service discovery over IPC channel e.g. message queue or UNIX domain socket is not performant since larger data is
+transferred, which can lead to transmission of several frames. If lots of services are discovered at high-frequency
+e.g. at startup the IPC channel can become a bottleneck. Furthermore, a lession learned from the iceoryx development
+so far is, that the IPC channel should just be used for the earliest communication during startup and not for the
+creation of objects inside the shared memory (see [this issue](https://github.com/eclipse-iceoryx/iceoryx/issues/611)).
+
+### Status quo in iceoryx Almond
+
+#### Runtime
+
+* `cxx::expected<InstanceContainer, FindServiceError> PoshRuntime::findService(capro::ServiceDescription)`
+  * Sends message over IPC channel
+
+#### RouDi
+
+* `runtime::IpcMessage {ProcessManager,PortManager}::findService(capro::ServiceDescription)`
+  * Called after RouDi has received a request from the runtime
+  * Sends back answer to `PoshRuntime` over IPC channel
+
+## Terminology
+
+## Requirements
+
+* Polling and event notification approach shall both be possible
+
+### Considerations and use-cases
+
+#### AUTOSAR Adaptive
+
+* findService() call as implemented in Almond v1.0
+* ara::com service discovery callbacks
+
+1. `StartFindService()`: Register callback, std::function interface
+1. `StopFindService()`: Unregister callback
+1. `FindService()`: one-shot, polling, synchronous
+
+#### SOME/IP-SD
+
+* Very similar to AUTOSAR Adaptive
+
+#### Cyclone DDS and ROS 2
+
+* Cyclone DDS calls this feature [topic discovery](https://github.com/eclipse-cyclonedds/cyclonedds/blob/master/docs/dev/topic_discovery.md). See also the [feature request](https://github.com/eclipse-cyclonedds/cyclonedds/issues/42).
+
+* Based on SDEP (simple endpoint discovery protocol)
+* `dds_ktopic`
+  * stores pointer to `ddsi_topic_definition`
+* Has built-in DCPSTopic topic and contains
+  * Topic name
+  * Type name
+  * Topic key
+  * QoS settings
+* `dds_find_topic_locally` function is available for users
+
+#### Static discovery
+
+* Static discovery shall be supported
+  * Having RouDi read in a config file, which has
+    * All the topics listed
+    * Wiring pub/sub information
+  * During init phase pub/sub are connected
+* Create abstract interface for `ServiceRegistry`
+  * Add new class to `StaticServiceRegistry::StaticServiceRegistry(std::map<CaproIdString_t, instance_t> map)`
+    * No notification-based callback possibilty
+    * Just static, polling with `FindService()` possible
+
+#### Sychronous, one-shot RPC (Polling)
+
+##### Alternative A: Request/ response feature usage
+
+* Re-use [request/ response feature](https://github.com/eclipse-iceoryx/iceoryx/issues/27)
+
+Pro:
+
+* Re-use building blocks to avoid code duplication
+* Dogfooding
+* Fast due to shared memory
+
+Con:
+
+* Shared memory approach might have overhead compared for a simple RPC call
```

Comment:
@elfenpiff @mossmaurice @elBoberido: How about this
1. We introduce a Discovery class for the public API 
2. This is a wrapper around a subscriber on a service registry topic with queue size 1
3. The class has a get() method that will pop the latest element from the queue, copies the data in an internal member and returns it
4. If you call the get() method and there is no new element in the queue, there is no change and you get the last service registry content again
5. You can add this class to a waitset or listener like a normal subscriber
6. We don't need a lot of chunks as all the queues will always hold the latest one and if it is consumed, a copy is made 
7. We have the same Discovery class with the same information for polling and event-based access

Reply:
I like the idea! I adapted alternative D accordingly and added some code examples. Another big plus from my point of view: why was `findService()` added to `PoshRuntime`? Because there was no other place to put it! Refactoring this out into a separate class, makes sense for a saner responsibility.